From 7b795534a39ad44d93ff1864ed98071c1f5bf6af Mon Sep 17 00:00:00 2001
Message-Id: <7b795534a39ad44d93ff1864ed98071c1f5bf6af@dist-git>
From: Michal Privoznik <mprivozn@redhat.com>
Date: Thu, 11 Apr 2019 15:14:18 -0400
Subject: [PATCH] qemu_hotplug: Attach guestfwd using netdev_add

https://bugzilla.redhat.com/show_bug.cgi?id=1624204

The guestfwd channels are -netdevs really. Hotplug them as such.

Signed-off-by: Michal Privoznik <mprivozn@redhat.com>
Reviewed-by: John Ferlan <jferlan@redhat.com>
(cherry picked from commit 903315dc8ffeb2af651207f090a53526decabc92)

Partially-Resolves: https://bugzilla.redhat.com/1658198
Signed-off-by: Laine Stump <laine@redhat.com>
Signed-off-by: Laine Stump <laine@laine.org>
Message-Id: <20190411191453.24055-7-laine@redhat.com>
Acked-by: Michal Privoznik <mprivozn@redhat.com>
---
 src/qemu/qemu_hotplug.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/src/qemu/qemu_hotplug.c b/src/qemu/qemu_hotplug.c
index 46cafef4c2..680c06c1ed 100644
--- a/src/qemu/qemu_hotplug.c
+++ b/src/qemu/qemu_hotplug.c
@@ -1753,10 +1753,14 @@ int qemuDomainAttachChrDevice(virQEMUDriverPtr driver,
     char *tlsAlias = NULL;
     const char *secAlias = NULL;
     bool need_release = false;
+    bool guestfwd = false;
 
-    if (chr->deviceType == VIR_DOMAIN_CHR_DEVICE_TYPE_CHANNEL &&
-        qemuDomainPrepareChannel(chr, priv->channelTargetDir) < 0)
-        goto cleanup;
+    if (chr->deviceType == VIR_DOMAIN_CHR_DEVICE_TYPE_CHANNEL) {
+        guestfwd = chr->targetType == VIR_DOMAIN_CHR_CHANNEL_TARGET_TYPE_GUESTFWD;
+
+        if (qemuDomainPrepareChannel(chr, priv->channelTargetDir) < 0)
+            goto cleanup;
+    }
 
     if (qemuAssignDeviceChrAlias(vmdef, chr, -1) < 0)
         goto cleanup;
@@ -1798,8 +1802,14 @@ int qemuDomainAttachChrDevice(virQEMUDriverPtr driver,
         goto exit_monitor;
     chardevAttached = true;
 
-    if (qemuMonitorAddDevice(priv->mon, devstr) < 0)
-        goto exit_monitor;
+    if (guestfwd) {
+        if (qemuMonitorAddNetdev(priv->mon, devstr,
+                                 NULL, NULL, 0, NULL, NULL, 0) < 0)
+            goto exit_monitor;
+    } else {
+        if (qemuMonitorAddDevice(priv->mon, devstr) < 0)
+            goto exit_monitor;
+    }
 
     if (qemuDomainObjExitMonitor(driver, vm) < 0)
         goto audit;
-- 
2.21.0

